lib/streamlit/runtime/secrets.py

```diff
@@ -150,6 +230,12 @@ def load_if_toml_exists(self) -> bool:
             # No secrets.toml files exist. That's fine.
             return False
 
+    def _add_provider(self, provider: SecretProvider):
```

Comment:
We have a couple of other interfaces that support custom implementations. Most (all?) of them are defined as part of `RuntimeConfig` and passed in when the Runtime is initialized:

https://github.com/streamlit/streamlit/blob/51cd0c91dfd9b2a0e90a147a4e09ce8bbc31d87f/lib/streamlit/runtime/runtime.py#L78C7-L78C20

Would it be an option to provide a list of secret providers when initializing the Runtime instead of via an `st.secrets. _add_provider` method?

Reply:
I considered this approach in this alternative branch:
https://github.com/streamlit/streamlit/commit/3cef9de12cbe4b2f169437d7201acc45e0592fbb#diff-fd8979106a82cdb83f15082f9e02b74e8d739964833a9980d5d74a745f6a77fbR128

this requires removing secret from being a singleton, which introduces more changes across the code. I am open to the idea of removing the singleton as part of this PR, but the original idea was to minimize the changes in this PR.

Reply:
+1 to what @LukasMasuch wrote. In context of this part in the PR description:
> adding internal st.secrets._add_provider() that will be used in SiS to wrap [UDF Style Secrets](https://docs.snowflake.com/en/developer-guide/external-network-access/secret-api-reference#python-api-for-secret-access).

I think it'd be better to make it an official way to reduce the chance of breaking anything since its not used in this code base here.

Reply:
I think it might be a welcome change and worth a discussion since we have been moving similar singletons to the runtime class. IMO, the ideal state would be that the runtime is the only singleton in Streamlit. If we are not doing this refactoring as part of this change, it might get a lot more complicated to do that in the future since SiS will depend on this `_add_provider` setup. I am linking @kajarenc and @raethlein for opinions since they have done some of these migrations in the past.

Reply:
@LukasMasuch  / @raethlein I can see the benefit of removing the singleton - i can add that to this PR too.

the simplest change to would be to make the constructor accept:
1. list of files from `secret.files` config, 2. `additional_providers` list.

But I should note that this doesn't seem ideal to me. This comes from the fact we don't let users define providers directly. A better design is a config lets users declare 1..n secret providers, and configuration for each provider. However I see that the current config is limited to single key/value (value can be array of simple values) but not additional nesting.

It will be beneficial to expand the config support to a single key to have for 1..n child config objects

And to be even more future proof, worth considering what I wrote in the PR with regard to "The provider in this PR only customize the behavior of _parse. So supporting credential providers such as UDF will require calling list secrets and then getting all secret values to populate the internal map"

Reply:
I can make a step towards that where `additional_providers` accepts `ProviderConfig` which will have provider name and config options. so in the future, providers implementation can be refactored without breaking changes.

Reply:
Would that solution keep `_add_provider`? I believe what Lukas and I are trying to say is that it would probably be best if [the RuntimeConfig](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/runtime.py#L78) has a new field `SecretProvider` that by default is set to the current behavior. This would then be identical to how you can init the runtime with a different cache_storage_manager, component_registry, media_file_storage etc.
When you initialize the runtime in SiS then, you can pass it a custom SecretProvider that is doing all the secret & file handling as required. In this case, you wouldn't need to have any new config options. In the future, we could think about allowing to state the SecretProvider module in the config or so, but it would be way more flexible than adding a `multiFileMode` now or similar while also following the pattern of the other configurable objects.

Reply:
Ah, I was thinking that having an `_add_provider` method in this case wouldn't be too unreasonable given that the way secrets currently work has them instantiated essentially right when the Streamlit library is imported.

I initially thought that we may want to keep things this way vs having a secrets provider set via `RuntimeConfig` so that we don't have to deal with a `*Proxy` class (like with session state / query params) to allow the user script to interact with a secrets provider.

I think I'm coming around to thinking that having secrets providers be configured differently from other pieces of Streamlit that can be given a custom implementation is likely a bad idea, though, and having to write a new `*Proxy` class is the better tradeoff to make.

Reply:
The one thing to keep in mind taking this approach is that this will mean that `st.secrets` can only be used in contexts where we have a Streamlit runtime initialized (kind of strange to think that we'd want to do this, but I believe I heard some discussion suggesting that we might).

Reply:
Yes.  I think we want secrets to work, even if the runtime is not initialized.  This is consistent with it being able to be used as an application framework that can be used for scripts that both run interctively, but can also be run in a non-connected enviornment.

Reply:
>  scripts that both run interctively, but can also be run in a non-connected environment.

@sfc-gh-wpugh Is there more information on the "non-connected environment" usage somewhere?